meta-refkit-core/recipes-ostree/initramfs-framework-ostree/files/ostree

```diff
@@ -29,27 +26,145 @@ ostree_debug () {
+            _ocurr=$_o
+        elif [ $_v -gt $_vprev ]; then
+            _vprev=$_v
+            _eprev=$_e
+            _oprev=$_o
         fi
     done
+ 
+    if [ -z "$_ecurr" ]; then
+        return 1
+    fi
 
-    [ -n "$WRAPPER" ] && return 0 || return 1
+    eval "${_evar}_CURR=\"$_ecurr\""
+    eval "${_evar}_PREV=\"$_eprev\""
+    eval "${_ovar}_CURR=\"${_ocurr#ostree=}\""
+    eval "${_ovar}_PREV=\"${_oprev#ostree=}\""
+
+    return 0
 }
 
-# Run the ostree image setup sequence.
-ostree_run () {
-    if ! ostree_discover_wrapper; then
-        ostree_fatal "could not find $SCRIPT in any deployment..."
+# Make the given directories/mount points movable.
+ostree_make_movable () {
+    for _d in $*; do
+        ostree_info "making $_d movable..."
+        mount --bind $_d $_d || return 1
+    done
+
+    return 0
+}
+
+# Fill in the missing details in an ostree deployment root.
+ostree_prepare_root () {
+    local _root=$1
+
+    #
+    # We need to prepare the ostree deployment root to be kosher for
+    # pivot-rooting and thus eventually becoming the final root.
+    # This involves stitching it together from several pieces to fill
+    # in the missing details. Currently we do this as follows:
+    #
+    #   - bind-mount the common /var into the deployment
+    #   - bind-mount the common /boot into the deployment
+    #   - bind-mount the common /home into the deployment
+
+    ostree_info "preparing $_root for pivot-rooting..."
+
+    cd $_root
+
+    mount --bind ../../var var || return 1
+    mount --bind /rootfs/boot $_root/boot || return 1
+    mount --bind /rootfs/home $_root/home || return 1
+
+    cd - > /dev/null
+
+    return 0
+}
+
+# Shuffle the ostree deployment around so that we eventually pivot-root it.
+ostree_shuffle_root () {
+    local _root="$1"
+
+    #
+    # This code mimicks the last bits of ostree-prepare-root. It shuffles
+    # /sysroot, /rootfs, and the chosen ostree deployment root ($_root) around
+    # so that /rootfs becomes $_root/sysroot and $_root becomes /rootfs. Then
+    # eventually after the final pivot-root we end up with
+    #
+    #   - current $_root (IOW the ostree deployment root) as /
+    #   - current /rootfs as /sysroot
+    #
+    # The only tricky/difficult to follow extra bit here is that we have to
+    # do these operations with an extra step to avoid trying to move the
+    # eventual /sysroot (now /rootfs) under itself.
+
+    ostree_info "shuffling /sysroot, /rootfs and $_root for pivot-rooting..."
+
+    cd $_root
+
+    # make /rootfs $_root/sysroot and $_root /rootfs
+    mkdir -p /sysroot.tmp || return 1
+    mount --move $_root /sysroot.tmp || return 1
+    mount --move /rootfs sysroot || return 1
+    mount --move . /rootfs || return 1
+
+    cd - > /dev/null
+
+    return 0
+}
+
+# Prepare the initramfs environment for pivot-rooting into an ostree deployment.
+ostree_initramfs_prepare_root () {
+    local _entry_CURR _entry_PREV _path_CURR _path_PREV
```

Comment:
@klihub it seems to me that this line doesn't actually do anything. These variables are set globally via the eval expressions in ostree_loader_entries(), so defining them as local in ostree_initramfs_prepare_root is redundant. It seems to work because "local" only declares them as local when defined, without actually defining the variables and thus shadowing the global ones - but I'm a bit fuzzy on the exact semantic, so my guess might be wrong.

Either way, this should get removed in a future cleanup commit. Agreed?